DESIGN.md

```diff
@@ -0,0 +1,400 @@
+    func (message *SignMessage) UnmarshalCBOR(data []byte) (err error)
+    func (m *SignMessage) Verify(external []byte, verifiers []Verifier) (err error)
+type Signature
+    func NewSignature() (s *Signature)
+    func (s *Signature) Decode(o interface{})
+    func (s *Signature) Equal(other *Signature) bool
+type Signer
+    func NewSigner(alg *Algorithm, options interface{}) (signer *Signer, err error)
+    func NewSignerFromKey(alg *Algorithm, privateKey crypto.PrivateKey) (signer *Signer, err error)
+    func (s Signer) GetAlg() *Algorithm
+    func (s *Signer) Public() (publicKey crypto.PublicKey)
+    func (s *Signer) Sign(rand io.Reader, digest []byte) (signature []byte, err error)
+    func (s *Signer) Verifier() (verifier *Verifier)
+type Verifier
+    func (v *Verifier) Verify(digest []byte, signature []byte) (err error)
+```
+
+Here are the proposed APIs:
+
+```go
+const (
+    HeaderLabelAlgorithm         = 1
+    HeaderLabelCritical          = 2
+    HeaderLabelContentType       = 3
+    HeaderLabelKeyID             = 4
+    HeaderLabelCounterSignature  = 7
+    HeaderLabelCounterSignature0 = 9
+    HeaderLabelX5Bag             = 32
+    HeaderLabelX5Chain           = 33
+    HeaderLabelX5T               = 34
+    HeaderLabelX5U               = 35
+)
+
+const (
+    CBORTagSignMessage  = 98
+    CBORTagSign1Message = 18
+)
+
+const (
+    ContextSignature  = "Signature"
+    ContextSignature1 = "Signature1"
+)
+
+var (
+    AlgorithmPS256   *Algorithm
+    AlgorithmPS384   *Algorithm
+    AlgorithmPS512   *Algorithm
+    AlgorithmES256   *Algorithm
+    AlgorithmES384   *Algorithm
+    AlgorithmES512   *Algorithm
+    AlgorithmEd25519 *Algorithm
+)
+
+type Algorithm struct {
+    Name     string
+    Value    int
+    HashFunc crypto.Hash
+}
+
+func I2OSP(x *big.Int, n int) ([]byte, error)
+func OS2IP(x []byte) *big.Int
+
+type ProtectedHeader map[interface{}]interface{}
+    func (h *ProtectedHeader) MarshalCBOR() ([]byte, error)
+    func (h *ProtectedHeader) UnmarshalCBOR(data []byte) error
+    func (h *ProtectedHeader) SetAlgorithm(alg *Algorithm)
+
+type UnprotectedHeader map[interface{}]interface{}
+
+type Headers struct {
+    RawProtected   []byte
+    RawUnprotected []byte
+    Protected      ProtectedHeader
+    Unprotected    UnprotectedHeader
+}
+
+type Signature struct {
+    Headers   Headers
+    Signature []byte
+}
+    func NewSignature() *Signature
+    func (s *Signature) MarshalCBOR() ([]byte, error)
+    func (s *Signature) UnmarshalCBOR(data []byte) error
+
+type SignMessage struct {
+    Headers    Headers
+    Payload    []byte
+    Signatures []*Signature
+}
+    func NewSignMessage() *SignMessage
+    func ParseSignMessage(data []byte) (*SignMessage, error)
+    func (m *SignMessage) MarshalCBOR() ([]byte, error)
+    func (m *SignMessage) UnmarshalCBOR(data []byte) error
+    func (m *SignMessage) Sign(rand io.Reader, external []byte, signers []Signer) error

```

Comment:
The more I use `SignMessage.Sign`, the more I feel like it is a leaky abstractions:
- It can only be called once, the second call will error out because the signed bytes have already been populated.
- `SignMessage.Signatures` and `signers` must have the same length.
- `SignMessage.Signatures` and `signers` algorithms must be in sync.

What if we define instead a top level function which creates the sane `SignMessage` and signs it:

```go
type Sig struct { // Find a  better name!
	Headers
	Signer
}

func Sign(rand io.Reader, headers Headers, payload, external []byte, signers []Sig) (*SignMessage, error)
```

Reply:
Let's do one iteration at a time. For the first iteration, we focus on the `COSE_Sign1`.